tests/Feature/Controllers/LabelControllerTest.php

```diff
@@ -0,0 +1,214 @@
+<?php
+
+namespace Tests\Feature\Controllers;
+
+use App\Contracts\AlbumRepositoryInterface;
+use CountriesSeeder;
+use App\Label;
+use App\Contracts\LabelRepositoryInterface;
+use App\Contracts\ArtistRepositoryInterface;
+use App\Contracts\CatalogEntityRepositoryInterface;
+use App\Contracts\UserRepositoryInterface;
+use App\Contracts\ProfileRepositoryInterface;
+#use App\Http\Resources\ArtistResource;
+#use App\Http\Resources\AlbumResource;
+
+class LabelControllerTest extends ControllerTestCase
+{
+    protected $label;
+
+    protected $artist;
+
+    protected $album;
+
+    protected $user;
+
+    protected $profile;
+
+    protected $catalogEntity;
+
+    /**
+     * @inheritdoc
+     */
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->seed(CountriesSeeder::class);
+
+        $this->label = resolve(LabelRepositoryInterface::class);
+
+        $this->artist = resolve(ArtistRepositoryInterface::class);
+
+        $this->album = resolve(AlbumRepositoryInterface::class);
+
+        $this->user = resolve(UserRepositoryInterface::class);
+
+        $this->profile = resolve(ProfileRepositoryInterface::class);
+
+        $this->catalogEntity = resolve(CatalogEntityRepositoryInterface::class);
+    }
+
+    protected function getExactJson(Label $model)
+    {
+        return [
+            'id' => $model->id,
+            'artists_count' => $model->artists->count(),
+            'albums_count' => $model->albums->count(),
+        ];
+    }
+
+    protected function getExactJsonWithArtistsAndAlbums(Label $model)
+    {
+        return [
+            'id' => $model->id,
+            // TODO These aren't necessary yet, but would look something like this.
+            #'artists' => ArtistResource::collection($model->artists),
+            #'albums' => AlbumResource::collection($model->albums),
+            'artists_count' => 1,
+            'albums_count' => 1,
+        ];
+    }
+
+    protected function getJsonStructureForLabel()
+    {
+        return [
+            'id',
+            'artists_count',
+            'albums_count',
+        ];
+    }
+
+    /**
+     * Ensure that a request for the index returns OK HTTP status and the
+     * expected JSON string.
+     */
+    public function testAllReturnsOkStatusAndExpectedJsonStructure()
+    {
+        $model = factory($this->label->class())->create();
+
+        $this->json('GET', route('labels.index'))
+            ->assertStatus(200)
+            ->assertExactJson([
+                'data' => [$this->getExactJson($model)]
+            ]);
+    }
+
+    /**
+     * Ensure that a request for an existing record returns OK HTTP status and
+     * the expected JSON string.
+     */
+    public function testShowReturnsOkStatusAndExpectedJsonStructure()
```

Comment:
Not sure that we need to assert the exact JSON string ... whats most important is just the structure .. Personally I wouldn't put in much effort comparing the actual values of the keys as its pointless in a test like this. Whats your thoughts ?

Reply:
I do think it's important that "the inputs match the outputs". This is easy to test in the way that we're doing it presently.

While testing for an exact JSON string match may be "redundant" in part, because the Repository tests overlap to some extent, I'm not sure it's "pointless".

I say this specifically because  HTTP (controller) tests introduce a number of other factors that could affect input/output equality. The `Resource` classes, for example, could potentially modify the output and we'd have no idea if we eliminate the exact JSON string match (because we're not testing Resources explicitly).

Even a mis-configured character set at the database level could butcher strings to the extent that the outputs do not match the inputs. How else do you propose we trap those types of problems?